test/e2e/framework/pods.go

```diff
@@ -33,33 +99,47 @@ func (f *Framework) WaitForPodToBeReady(waitedPod *v1.Pod, cluster ClusterIndex)
 		return true, nil // pods is running
 	})
 	Expect(err).NotTo(HaveOccurred())
-	return finalPod
+	tp.Pod = finalPod
 }
 
-func (f *Framework) WaitForPodFinishStatus(waitedPod *v1.Pod, cluster ClusterIndex) (terminationCode int32, terminationMessage string) {
-	pc := f.ClusterClients[cluster].CoreV1().Pods(f.Namespace)
+func (tp *TestPod) WaitForFinishStatus() {
+
+	finished := tp.WaitForFinish()
+
+	Expect(finished).To(BeTrue())
+
+	tp.TerminationCode = tp.Pod.Status.ContainerStatuses[0].State.Terminated.ExitCode
+	tp.TerminationMessage = tp.Pod.Status.ContainerStatuses[0].State.Terminated.Message
+}
+
+func (tp *TestPod) WaitForFinish() bool {
+	pc := tp.framework.ClusterClients[tp.Config.Cluster].CoreV1().Pods(tp.framework.Namespace)
+
 	err := wait.PollImmediate(5*time.Second, 2*time.Minute, func() (bool, error) {
-		pod, err := pc.Get(waitedPod.Name, metav1.GetOptions{})
+		var err error
+		tp.Pod, err = pc.Get(tp.Pod.Name, metav1.GetOptions{})
 		if err != nil {
 			if IsTransientError(err) {
 				Logf("Transient failure when attempting to list pods: %v", err)
 				return false, nil // return nil to avoid PollImmediate from stopping
 			}
 			return false, err
 		}
-		switch pod.Status.Phase {
+		switch tp.Pod.Status.Phase {
 		case v1.PodSucceeded:
-			terminationCode = pod.Status.ContainerStatuses[0].State.Terminated.ExitCode
-			terminationMessage = pod.Status.ContainerStatuses[0].State.Terminated.Message
 			return true, nil
 		case v1.PodFailed:
-			terminationCode = pod.Status.ContainerStatuses[0].State.Terminated.ExitCode
-			terminationMessage = pod.Status.ContainerStatuses[0].State.Terminated.Message
 			return true, nil
 		default:
 			return false, nil
 		}
 	})
+
 	Expect(err).NotTo(HaveOccurred())
-	return terminationCode, terminationMessage
+
+	return tp.Pod.Status.Phase == v1.PodSucceeded || tp.Pod.Status.Phase == v1.PodFailed
```

Comment:
Wouldn't it have a different status only when it had an error (based on the switch on line 128)? In that case it would fail on line 138 anyway, making this check moot

Reply:
That one (L128) returns to the inner loop of wait.PollInmediate

You could have still existed while the pod is pending, due to a timeout. In that case I want it to return "false",
because the Pod didn't really "Finish" and we are waiting for finish status.

Reply:
I reckon that when timeout is reached it will return an error, in fact "ErrWaitTimeout is returned when the condition exited without success." from https://godoc.org/k8s.io/apimachinery/pkg/util/wait#pkg-variables